### ISSUE TITLE
Incomplete call hierarchy

### ISSUE THREAD

**Issue by fedejeanne** [vi-eclipse/Eclipse-JDT#9]
The _call hierarchy_ is sometimes incomplete. This might have something to do with https://github.com/vi-eclipse/Eclipse-JDT/issues/4 but the workaround proposed there did **not** help.

The problem is reproducible in Trunk, just look for the call hierarchy (`CTRL` + `ALT` + `h`) of: 

- `ru.novosoft.mdf.impl.modelcache.persistence.file.AbstractMappingHandler.streamMappings(OutputStream, Collection<Pair<String, String>>)`:

---

![image](https://github.com/vi-eclipse/Eclipse-JDT/assets/2205684/3bdb1198-5a4a-44ff-bf8e-58a72e3f675a)

---

It looks like the method is not being called but doing a "manual renaming" (i.e. without refactoring) reveals some calls:

---

![Image](https://user-images.githubusercontent.com/2205684/256267142-f4c5dca6-8a6c-4ad1-a4b5-5212424ddbd6.png)

**Comment by fedejeanne** [comment]
The problem is not reproducible in `2023-06`

**Comment by fedejeanne** [comment]
The problem is reproducible in `2023-06` but not in `2023-09 M1` (at least for now). I'll leave this issue open until `2023-09` is released and rolled out as out "stable" IDE and try to reproduce it one more time.

If this error is only sporadically reproducible then it probably is a consequence of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1303

**Comment by HeikoKlare** [comment]
This issue seems to be related or even the same: https://bugs.eclipse.org/bugs/show_bug.cgi?id=552774
It provides multiple reproduction scenarios. Some cases where caller are missing have been addressed in 4.28 (which conforms to the results of your analysis, @fedejeanne). But the issue does not seem to be fixed completely, so we may have another look we the latest reproduction scenario provided by Eike in that bug report.

**Comment by fedejeanne** [comment]
Thank you @HeikoKlare , the issue is now reproducible in `2023-09`.

I will look into it.

**Comment by fedejeanne** [comment]
Happened again in **4.37.0.v20250804-1800** (M3) for the method: 
- `AdvancedPopulateHelper.getAllLegendDiagramForAdvancedPopulate(RefObject)`

**Comment by fedejeanne** [comment]
## TL;DR

The problem is that the deep hierarchy of classes (some in JAR files and some in the WS) produce an **inaccurate** match and therefore the matches are not shown in the UI.

## Long answer

The method accepts a `RefObject` as a parameter but the callers pass a subclass (`MDIRepresentableArtefact`) that is 7 levels deep in the hierarchy. Here's the hierarchy, I've marked the classes with _jar_ to indicate they are present in a JAR file in the classpath or with _src_ to indicate that they are present as source code in the WS:

`MDIRepresentableArtefact` _(jar)_ -> `MHasVariantAttributes` _(jar)_ -> `MCoreArtefact` _(jar)_ -> `MAbstractMementoOwner` _(jar)_ -> `MArtefact` _(jar)_ -> `MDFObject` (src) -> `RefObject` (src)

If I change the type of the parameter in the caller and I set it explicitly to `RefObject` then I can see the caller in the result:

<img width="908" height="305" alt="Image" src="https://github.com/user-attachments/assets/657b14e0-ee9f-4b91-89fe-3aff3f54bc61" />

---

But if I change it to anything else in the hierarchy then I can **not** see the match (I've tried each one of the classes mentioned before).

Last but not least, the callers that I would like to see in the call hierarchy are all correctly located by the `JavaSearchParticipant`, but debugging I noticed that when the matches reach [this line](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/e46bffe62303510a56d246184d6d570349ab1029/org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/MatchLocator.java#L2212) in `org.eclipse.jdt.internal.core.search.matching.MatchLocator.report(SearchMatch)`:

```java
this.requestor.acceptSearchMatch(match);
```

... the `match` has an `accuracy` of `1` (_a.k.a._ `INACCURATE`). If I manually change the accuracy to `0` (_a.k.a._ `ACCURATE`) then I see the matches in the call hierarchy view.

**Comment by fedejeanne** [comment]
@HeikoKlare didn't we have other issues that were also related to MM classes being present in the JAR files? This problem rings a bell, I just can't quite pinpoint it.

**Comment by HeikoKlare** [comment]
We have some problems with different kinds of Java tooling (class hierarchy with method overwrites, conditional breakpoints etc.) because of our structure of workspace and TP bundles: workspace bundles depend on TP bundles which, in turn, depend on workspace bundles. If I remember correctly, the TP bundles get some fake "assignment" to the first workspace bundle that has a dependency to it, which leads to some problems (in particular when coming from other bundles having a dependency to them as well).
Andreas has already analyzed this topic with a student and another colleague, so he probably knows more about the details.

**Comment by fedejeanne** [comment]
FTR, I learned that the problem is an incorrect `TypeBinding` that believes that `MDFObject` does **not** extend from `RefObject`

<img width="881" height="411" alt="Image" src="https://github.com/user-attachments/assets/ea133482-1ff4-40f5-94dd-87575f7a8f35" />

When this happens, the call is not "compatible" with the method definition and therefore the match is `INACCURATE`. More precisely, the method `org.eclipse.jdt.internal.compiler.lookup.Scope.findMethod(ReferenceBinding, char[], TypeBinding[], InvocationSite, boolean)` will return an **invalid** binding (a `ProblemMethodBinding`) in line 3073 and this will end up messing up the results

<img width="863" height="533" alt="Image" src="https://github.com/user-attachments/assets/4d16893e-474e-46d1-91ac-88c1fb0912ea" />

---

<details><summary>Stack-trace where the parameter comparison goes wrong</summary>
<p>

```java
Thread [Worker-66: Fetching children of getAllLegendDiagramForAdvancedPopulate] (Suspended)	
	MethodScope(Scope).parameterCompatibilityLevel(TypeBinding, TypeBinding, LookupEnvironment, MethodBinding) line: 4983	
	MethodScope(Scope).parameterCompatibilityLevel(MethodBinding, TypeBinding[], InvocationSite) line: 4931	
	MethodScope(Scope).computeCompatibleMethod(MethodBinding, TypeBinding[], InvocationSite) line: 833	
	MethodScope(Scope).findMethod0(ReferenceBinding, char[], TypeBinding[], InvocationSite, boolean) line: 1721	
	MethodScope(Scope).findMethod(ReferenceBinding, char[], TypeBinding[], InvocationSite, boolean) line: 1616	
	MethodScope(Scope).getMethod(TypeBinding, char[], TypeBinding[], InvocationSite) line: 3073	
	MessageSend.findMethodBinding(BlockScope) line: 1128	
	MessageSend.resolveType(BlockScope) line: 870	
	LocalDeclaration.resolve(BlockScope) line: 332	
	LocalDeclaration(Statement).resolveWithBindings(LocalVariableBinding[], BlockScope) line: 523	
	ASTNode.resolveStatements(Statement[], BlockScope) line: 696	
	MethodDeclaration(AbstractMethodDeclaration).resolveStatements() line: 725	
	MethodDeclaration.resolveStatements() line: 373	
	MethodDeclaration(AbstractMethodDeclaration).resolve(ClassScope) line: 624	
	TypeDeclaration.resolve() line: 1442	
	TypeDeclaration.resolve(CompilationUnitScope) line: 1571	
	CompilationUnitDeclaration.resolve() line: 661	
	MatchLocator.process(PossibleMatch, boolean) line: 2051	
	MatchLocator.locateMatchesDefaultImpl(JavaProject, PossibleMatch[], int, int) line: 1360	
	MatchLocator.locateMatches(JavaProject, PossibleMatch[], int, int) line: 1279	
	MatchLocator.locateMatches(JavaProject, PossibleMatchSet, int) line: 1397	
	MatchLocator.locateMatches(SearchDocument[]) line: 1518	
	JavaSearchParticipant.locateMatches(SearchDocument[], SearchPattern, IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 143	
	BasicSearchEngine.findMatches(SearchPattern, SearchParticipant[], IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 278	
	BasicSearchEngine.search(SearchPattern, SearchParticipant[], IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 622	
	SearchEngine.search(SearchPattern, SearchParticipant[], IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 677	
	CallerMethodWrapper.findChildren(IProgressMonitor) line: 155	
	CallerMethodWrapper(MethodWrapper).performSearch(IProgressMonitor) line: 265	
	CallerMethodWrapper(MethodWrapper).doFindChildren(IProgressMonitor) line: 196	
	CallerMethodWrapper(MethodWrapper).getCalls(IProgressMonitor) line: 87	
	DeferredMethodWrapper.getCalls(IProgressMonitor) line: 72	
	DeferredMethodWrapper.lambda$0(IElementCollector, DeferredMethodWrapper, IProgressMonitor) line: 80	
	Lambda.run() line: not available	
	JavaCore.lambda$1(JavaCore$JavaRunnable) line: 6178	
	Lambda.call() line: not available	
	JavaModelManager.cacheZipFiles(JavaCallable<T,E>) line: 5703	
	JavaModelManager.callReadOnly(JavaCallable<T,E>) line: 5692	
	JavaCore.callReadOnly(JavaCallable<T,E>) line: 6162	
	JavaCore.runReadOnly(JavaRunnable<E>) line: 6177	
	DeferredMethodWrapper.fetchDeferredChildren(Object, IElementCollector, IProgressMonitor) line: 80	
	DeferredTreeContentManager$1.run(IProgressMonitor) line: 215	
	Worker.run() line: 63	
```


</p>
</details>

**Comment by fedejeanne** [comment]
I found the line of code where things go wrong: it's in [`org.eclipse.jdt.internal.compiler.lookup.ClassScope.detectHierarchyCycle(SourceTypeBinding, ReferenceBinding, TypeReference)`](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/e9f2f833e6dcc009cd25a900efe614710314869c/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ClassScope.java#L1599). In a nutshell, our extremely complicated hierarchy of interfaces **with annotations** in our metamodel "has a cycle" (allegedly) .

The cycle looks like this:

1. `A` extends `B`
2. `B` extends `C`
3. `C` extends `D`
4. `D` extends `E` and `L`
5. `E` extends `F`
6. `F` extends `G`
7. `G` (extends nothing,) has annotations and imports `H`
8. `H` extends `I`
9. `I` extends `G`, has annotations and imports `J`
10. `J` extends `H` and `K<L>` (yes, generics are also part of the mix)
11. `L` extends `E`

The cycle is reported while resolving `L` and the problem (an instance of `DefaultProblem` that is created [here](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/e9f2f833e6dcc009cd25a900efe614710314869c/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/problem/ProblemHandler.java#L172-L183)) looks like this:

```java
Pb(317) Cycle detected: a cycle exists in the type hierarchy between L and E
```

Here's the stack trace right before reporting the issue:

<details><summary>Stack trace</summary>
<p>

```java
Thread [Worker-44: Fetching children of getAllLegendDiagramForAdvancedPopulate] (Suspended (breakpoint at line 1596 in ClassScope))	
	ClassScope.detectHierarchyCycle(SourceTypeBinding, ReferenceBinding, TypeReference) line: 1596	
	ClassScope.detectHierarchyCycle(TypeBinding, TypeReference) line: 1492	
	SingleTypeReference.getTypeBinding(Scope) line: 76	
	SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	SingleTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	SingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	ClassScope(Scope).getTypeOrPackage(char[], int, boolean) line: 3500	
	ClassScope(Scope).getType(char[]) line: 3202	
	SingleTypeReference.getTypeBinding(Scope) line: 60	
	SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	SingleTypeReference(TypeReference).resolveTypeArgument(ClassScope, ReferenceBinding, int) line: 639	
	ParameterizedSingleTypeReference.internalResolveLeafType(Scope, ReferenceBinding, boolean) line: 262	
	ParameterizedSingleTypeReference.internalResolveType(Scope, ReferenceBinding, boolean, int) line: 173	
	ParameterizedSingleTypeReference.resolveType(ClassScope, int) line: 391	
	ParameterizedSingleTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	ParameterizedSingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	CompilationUnitScope.findImport(char[][], int) line: 626	
	CompilationUnitScope.findSingleImport(char[][], int, boolean) line: 693	
	CompilationUnitScope.faultInImports() line: 525	
	MethodScope(Scope).getBinding(char[], int, InvocationSite, boolean) line: 2190	
	MethodScope(BlockScope).getBinding(char[][], int, InvocationSite, boolean) line: 493	
	QualifiedNameReference.resolveType(BlockScope) line: 1007	
	MemberValuePair.resolveTypeExpecting(BlockScope, TypeBinding) line: 77	
	SingleMemberAnnotation(Annotation).resolveType(BlockScope) line: 959	
	ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding, boolean) line: 899	
	ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding) line: 790	
	SourceTypeBinding.getAnnotationTagBits() line: 1143	
	TypeDeclaration.updateSupertypesWithAnnotations(Map<ReferenceBinding,ReferenceBinding>) line: 1837	
	CompilationUnitScope.integrateAnnotationsInHierarchy() line: 426	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 194	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	CompilationUnitScope.findImport(char[][], int) line: 626	
	CompilationUnitScope.findSingleImport(char[][], int, boolean) line: 693	
	CompilationUnitScope.resolveSingleImport(ImportBinding, int) line: 954	
	ClassScope(Scope).getTypeOrPackage(char[], int, boolean) line: 3480	
	ClassScope(Scope).getType(char[]) line: 3202	
	SingleTypeReference.getTypeBinding(Scope) line: 60	
	SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	SingleTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	SingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	CompilationUnitScope.findImport(char[][], int) line: 626	
	CompilationUnitScope.findSingleImport(char[][], int, boolean) line: 693	
	CompilationUnitScope.faultInImports() line: 525	
	MethodScope(Scope).getBinding(char[], int, InvocationSite, boolean) line: 2190	
	MethodScope(BlockScope).getBinding(char[][], int, InvocationSite, boolean) line: 493	
	QualifiedNameReference.resolveType(BlockScope) line: 1007	
	MemberValuePair.resolveTypeExpecting(BlockScope, TypeBinding) line: 77	
	SingleMemberAnnotation(Annotation).resolveType(BlockScope) line: 959	
	ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding, boolean) line: 899	
	ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding) line: 790	
	SourceTypeBinding.getAnnotationTagBits() line: 1143	
	TypeDeclaration.updateSupertypesWithAnnotations(Map<ReferenceBinding,ReferenceBinding>) line: 1837	
	CompilationUnitScope.integrateAnnotationsInHierarchy() line: 426	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 194	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	ClassScope(Scope).getTypeOrPackage(char[], int, boolean) line: 3500	
	ClassScope(Scope).getType(char[]) line: 3202	
	SingleTypeReference.getTypeBinding(Scope) line: 60	
	SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	SingleTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	SingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	ClassScope(Scope).getTypeOrPackage(char[], int, boolean) line: 3500	
	ClassScope(Scope).getType(char[]) line: 3202	
	SingleTypeReference.getTypeBinding(Scope) line: 60	
	SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	SingleTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	SingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	CompilationUnitScope.findImport(char[][], int) line: 626	
	CompilationUnitScope.findSingleImport(char[][], int, boolean) line: 693	
	CompilationUnitScope.resolveSingleImport(ImportBinding, int) line: 954	
	ClassScope(Scope).getTypeOrPackage(char[], int, boolean) line: 3480	
	ClassScope(Scope).getType(char[]) line: 3202	
	SingleTypeReference.getTypeBinding(Scope) line: 60	
	SingleTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	SingleTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	SingleTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	SingleTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	ClassScope(Scope).getPackage(char[][]) line: 3119	
	QualifiedTypeReference.getTypeBinding(Scope) line: 111	
	QualifiedTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	QualifiedTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	QualifiedTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	QualifiedTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	ClassScope(Scope).getPackage(char[][]) line: 3119	
	QualifiedTypeReference.getTypeBinding(Scope) line: 111	
	QualifiedTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	QualifiedTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	QualifiedTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	QualifiedTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	ClassScope(Scope).getPackage(char[][]) line: 3119	
	QualifiedTypeReference.getTypeBinding(Scope) line: 111	
	QualifiedTypeReference(TypeReference).internalResolveType(Scope, int) line: 509	
	QualifiedTypeReference(TypeReference).resolveType(ClassScope, int) line: 621	
	QualifiedTypeReference(TypeReference).resolveType(ClassScope) line: 617	
	QualifiedTypeReference(TypeReference).resolveSuperType(ClassScope) line: 589	
	ClassScope.findSupertype(TypeReference) line: 1622	
	ClassScope.connectSuperInterfaces() line: 1321	
	ClassScope.connectTypeHierarchy() line: 1430	
	CompilationUnitScope.connectTypeHierarchy() line: 408	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 190	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	CompilationUnitScope.findImport(char[][], int) line: 626	
	CompilationUnitScope.findSingleImport(char[][], int, boolean) line: 693	
	CompilationUnitScope.faultInImports() line: 525	
	MethodScope(Scope).getBinding(char[], int, InvocationSite, boolean) line: 2190	
	MethodScope(BlockScope).getBinding(char[][], int, InvocationSite, boolean) line: 493	
	QualifiedNameReference.resolveType(BlockScope) line: 1007	
	MemberValuePair.resolveTypeExpecting(BlockScope, TypeBinding) line: 77	
	SingleMemberAnnotation(Annotation).resolveType(BlockScope) line: 959	
	ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding, boolean) line: 899	
	ASTNode.resolveAnnotations(BlockScope, Annotation[], Binding) line: 790	
	SourceTypeBinding.getAnnotationTagBits() line: 1143	
	TypeDeclaration.updateSupertypesWithAnnotations(Map<ReferenceBinding,ReferenceBinding>) line: 1837	
	CompilationUnitScope.integrateAnnotationsInHierarchy() line: 426	
	LookupEnvironment$CompleteTypeBindingsSteps.perform(CompilationUnitScope) line: 194	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean, boolean) line: 629	
	LookupEnvironment.completeTypeBindings(CompilationUnitDeclaration, boolean) line: 621	
	MatchLocator.accept(ICompilationUnit, AccessRestriction) line: 369	
	LookupEnvironment.askForType(PackageBinding, char[], ModuleBinding) line: 382	
	PlainPackageBinding(PackageBinding).getTypeOrPackage(char[], ModuleBinding, boolean) line: 274	
	CompilationUnitScope.findImport(char[][], int) line: 626	
	CompilationUnitScope.findSingleImport(char[][], int, boolean) line: 693	
	CompilationUnitScope.faultInImports() line: 525	
	CompilationUnitScope.faultInTypes() line: 600	
	MatchLocator.process(PossibleMatch, boolean) line: 2049	
	MatchLocator.locateMatchesDefaultImpl(JavaProject, PossibleMatch[], int, int) line: 1360	
	MatchLocator.locateMatches(JavaProject, PossibleMatch[], int, int) line: 1279	
	MatchLocator.locateMatches(JavaProject, PossibleMatchSet, int) line: 1397	
	MatchLocator.locateMatches(SearchDocument[]) line: 1518	
	JavaSearchParticipant.locateMatches(SearchDocument[], SearchPattern, IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 143	
	BasicSearchEngine.findMatches(SearchPattern, SearchParticipant[], IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 278	
	BasicSearchEngine.search(SearchPattern, SearchParticipant[], IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 622	
	SearchEngine.search(SearchPattern, SearchParticipant[], IJavaSearchScope, SearchRequestor, IProgressMonitor) line: 677	
	CallerMethodWrapper.findChildren(IProgressMonitor) line: 155	
	CallerMethodWrapper(MethodWrapper).performSearch(IProgressMonitor) line: 265	
	CallerMethodWrapper(MethodWrapper).doFindChildren(IProgressMonitor) line: 196	
	CallerMethodWrapper(MethodWrapper).getCalls(IProgressMonitor) line: 87	
	DeferredMethodWrapper.getCalls(IProgressMonitor) line: 72	
	DeferredMethodWrapper.lambda$0(IElementCollector, DeferredMethodWrapper, IProgressMonitor) line: 80	
	Lambda.run() line: not available	
	JavaCore.lambda$1(JavaCore$JavaRunnable) line: 6178	
	Lambda.call() line: not available	
	JavaModelManager.cacheZipFiles(JavaCallable<T,E>) line: 5703	
	JavaModelManager.callReadOnly(JavaCallable<T,E>) line: 5692	
	JavaCore.callReadOnly(JavaCallable<T,E>) line: 6162	
	JavaCore.runReadOnly(JavaRunnable<E>) line: 6177	
	DeferredMethodWrapper.fetchDeferredChildren(Object, IElementCollector, IProgressMonitor) line: 80	
	DeferredTreeContentManager$1.run(IProgressMonitor) line: 215	
	Worker.run() line: 63	
```

</p>
</details> 

I'm working on a reproducer.

**Comment by fedejeanne** [comment]
FTR I am testing it again in `2025-09` and it is **not** reproducible. I tested with these 2 methods:

- `ru.novosoft.mdf.impl.modelcache.persistence.file.AbstractMappingHandler.streamMappings(OutputStream, Collection<Pair<String, String>>)`
- `aquintos.gdf.advancedpopulate.AdvancedPopulateHelper.getAllLegendDiagramForAdvancedPopulate(RefObject)`

I see no evidence of problems with the index in the trace logs either.

**Comment by fedejeanne** [comment]
I tested again today in the latest I-BUILD (**4.38.0.v20250926-0030**) and it **is** reproducible, this time only for `aquintos.gdf.advancedpopulate.AdvancedPopulateHelper.getAllLegendDiagramForAdvancedPopulate(RefObject)`.

I tested before and after rebuilding the indexes: there were no errors while building the indexes (https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4388) and the result was the same so I guess our initial assumption that this issue might be related to #4 is wrong (or at least, partially wrong).

**Comment by fedejeanne** [comment]
Reported in: 
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2575

Fixed in **2025-12** by:
- https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4526
